src/events/nats.js

```diff
@@ -0,0 +1,170 @@
+            }
+            const { slsEvent: natsPublishEvent, startTime } = eventInterface.initializeEvent(
+                NATS_TYPES.name,
+                serverHostname,
+                'publish',
+                NATS_TYPES.name
+            );
+            const responseMetadata = {
+                subject: subject_internal,
+            };
+            if (!jsonConnectProperty) {
+                responseMetadata.msg = msg_internal;
+            } else if (msgJsonStringify && msgJsonStringify !== NATS_TYPES.badMessage) {
+                responseMetadata.msg = msgJsonStringify;
+            }
+            let patchedCallback = opt_callback_internal;
+
+            if (opt_callback_internal) {
+                const promise = new Promise((resolve) => {
+                    patchedCallback = () => {
+                        eventInterface.finalizeEvent(
+                            natsPublishEvent,
+                            startTime,
+                            null,
+                            responseMetadata
+                        );
+                        resolve();
+                        opt_callback_internal();
+                    };
+                });
+                tracer.addEvent(natsPublishEvent, promise);
+            } else {
+                tracer.addEvent(natsPublishEvent);
+                eventInterface.finalizeEvent(
+                    natsPublishEvent,
+                    startTime,
+                    null,
+                    responseMetadata
+                );
+            }
+            return original.apply(this,
+                [subject_internal, msg_internal, opt_reply_internal, patchedCallback]);
+        } catch (err) {
+            tracer.addException(err);
+            if (!clientRequest) {
+                clientRequest = original.apply(this,
+                    [subject, msg, opt_reply, opt_callback]);
+            }
+        }
+        return clientRequest;
+    };
+}
+
+/**
+ * Wrap nats connect function.
+ * @param {Function} connectFunction nats connect function.
+ * @returns {Function} The wrapped function
+ */
+function wrapNatsConnectFunction(connectFunction) {
+    return function internalNatsConnectFunction(url, opts) {
+        const connectFunctionResponse = connectFunction(url, opts);
+        if (connectFunctionResponse && connectFunctionResponse.constructor) {
+            if (connectFunctionResponse.constructor.name !== NATS_TYPES.mainWrappedFunction) {
+                return connectFunctionResponse;
+            }
+            const serverHostname = getServerHostname(connectFunctionResponse.currentServer);
+            const jsonConnectProperty = connectFunctionResponse.options ?
+                connectFunctionResponse.options.json : null;
+            shimmer.wrap(connectFunctionResponse, 'publish', () => wrapNatsPublishFunction(connectFunctionResponse.publish, serverHostname, jsonConnectProperty));
+        }
```

Comment:
maybe wrap all of this with a try/catch block

Reply:
ok